### PR TITLE
Add perf counters

### PR DESCRIPTION
**Pull request by pmaytak**
Added cache related counters: cache size, read misses, read and write durations.
Refactored BenchmarkMsalDistributedTokenCacheAdapter, BenchmarkMsalMemoryTokenCacheProvider to directly call the caches because the old way (making these classes a wrapper) was breaking up the method signatures.
Other minor refactorings.

Issue #643 partially.

### CODE REVIEW

File: tests/PerformanceTests/PerformanceTestService/EventSource/BenchmarkMsalDistributedTokenCacheAdapter.cs [hunk 1]
```diff
@@ -25,10 +26,17 @@ public class BenchmarkMsalDistributedTokenCacheAdapter : MsalDistributedTokenCac
...
         /// <returns>A <see cref="Task"/> that completes when key removal has completed.</returns>
-        protected override async Task RemoveKeyAsync(string cacheKey)
+        protected override Task RemoveKeyAsync(string cacheKey)
         {
+            var bytes = base.ReadCacheBytesAsync(cacheKey).GetAwaiter().GetResult();
+
+            if (bytes != null)
+            {
+                MemoryCacheEventSource.Log.DecrementSize(bytes.Length);
+            }
```
Comment (henrik-me) [inline tests/PerformanceTests/PerformanceTestService/EventSource/BenchmarkMsalDistributedTokenCacheAdapter.cs:36]: what is the cpu cost of doing this?

Comment (pmaytak) [reply tests/PerformanceTests/PerformanceTestService/EventSource/BenchmarkMsalDistributedTokenCacheAdapter.cs:36]: As per docs, EventSource is low overhead. AppInsights SDK gets the data every 60 seconds. If we want to collect periodically, there's no way to directly get item count and size from the memory caches, so seems like we would still have to have some sort of a counter running.

File: tests/PerformanceTests/PerformanceTestService/EventSource/MemoryCacheEventSource.cs [hunk 2]
```diff
@@ -54,32 +63,80 @@ public void IncrementReadCount()
...
+        /// <remarks>
+        /// This is the convention for initializing counters in the RuntimeEventSource (lazily on the first enable command).
+        /// They aren't disabled afterwards
+        /// </remarks>
         protected override void OnEventCommand(EventCommandEventArgs command)
         {
             if (command.Command == EventCommand.Enable)
             {
-                _cacheItemCounter ??= new PollingCounter(CacheItemCounterName, this, () => _totalItemCount)
+                _cacheItemCounter ??= new PollingCounter(CacheItemCounterName, this, () => Volatile.Read(ref _totalItemCount))
```
Comment (gfoidl) [inline tests/PerformanceTests/PerformanceTestService/EventSource/MemoryCacheEventSource.cs:105]: FYI there was a discussion whether `Volatile.Read` or `Interlocked.Read` should be used.
The [consensus](https://github.com/dotnet/docs/pull/20984#discussion_r502028017) was to use `Interlocked.Read` in this case.

Comment (pmaytak) [reply tests/PerformanceTests/PerformanceTestService/EventSource/MemoryCacheEventSource.cs:105]: Awesome, thanks @gfoidl. This is very helpful.

File: tests/PerformanceTests/PerformanceTestService/EventSource/BenchmarkMsalDistributedTokenCacheAdapter.cs [hunk 3]
```diff
@@ -40,8 +72,18 @@ protected override async Task RemoveKeyAsync(string cacheKey)
         /// (account or app).</returns>
         protected override async Task<byte[]> ReadCacheBytesAsync(string cacheKey)
         {
+            var stopwatch = Stopwatch.StartNew();
+            var bytes = await _distributedCache.GetAsync(cacheKey).ConfigureAwait(false);
+            stopwatch.Stop();
+
             MemoryCacheEventSource.Log.IncrementReadCount();
-            return await base.ReadCacheBytesAsync(cacheKey);
+            MemoryCacheEventSource.Log.AddReadDuration(stopwatch.Elapsed.TotalMilliseconds);
```
Comment (jmprieur) [inline tests/PerformanceTests/PerformanceTestService/EventSource/BenchmarkMsalDistributedTokenCacheAdapter.cs:80]: Is it to measure the time of the the distributed store read?

**Review by jmprieur — APPROVED** [review]
LGTM.
Thanks @pmaytak

File: tests/PerformanceTests/PerformanceTestService/EventSource/BenchmarkMsalDistributedTokenCacheAdapter.cs (hunk 3 above)
Comment (pmaytak) [reply tests/PerformanceTests/PerformanceTestService/EventSource/BenchmarkMsalDistributedTokenCacheAdapter.cs:80]: Yes.

**Pull request merged**